README.md
AWS Least Privilege
===================

[![npm package][npm-image]][npm-url] 
[![Build Status][travis-image]][travis-url] 
[![Coverage Status][coveralls-image]][coveralls-url] 
[![Dependencies Status][david-image]][david-url]

Use AWS X-Ray to reach Least Privilege.

This project aims to streamline the process of collecting resource usage information from X-Ray and reaching a "Least Privilege" security posture for a given application. AWS X-Ray provides in-depth information about service API calls executed via the AWS SDK. Using this information, it is possible to build a profile of the AWS resources and actions that are actually used by an application and generate a policy document reflecting it.  The project is currently focused on AWS Lambda but can easily be applied to other applications that utilize AWS Roles (applications on EC2 or ECS).

# Requirements

NodeJS 6+

# Installation

```
npm install -g aws-least-privilege
```

This will install the command line tool: `xray-privilege-scan`.

# Credential Setup

The cli tool uses AWS Nodejs SDK internally and will use the same credential mechanism as used by the SDK. It will automatically use credentials from the AWS shared credential file. See: [AWS SDK Docs](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-shared.html) for more details. The user used to run the cli should have the AWS managed policy: `AWSXrayReadOnlyAccess`. 

If using compare mode (see below) then the following inline policy should be attached to the user:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iam:ListAttachedRolePolicies",
                "iam:ListRolePolicies",
                "iam:GetPolicy",
                "iam:GetPolicyVersion",
                "iam:GetRolePolicy",
                "lambda:GetFunctionConfiguration"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

# X-Ray Setup

Follow the [AWS X-Ray Developer Guide](https://docs.aws.amazon.com/xray/latest/devguide/aws-xray.html) to install and enable X-Ray for your application.

## X-Ray Parameter Whitelist Configuration

In order to capture what resources are accessed by AWS service calls, AWS X-Ray needs to be configured to send out parameter information as part of the subsegment traces. For example for DynamoDB we would want to know the Table being accessed and for S3 the Bucket and Key. AWS X-Ray currently provides default parameter info only for a limited set of services: 

* DynamoDB: All X-Ray SDKs
* SQS: All X-Ray SDKs
* Lambda: All X-Ray SDKs 
* S3: 
  * Node SDK: version 1.2.0 and higher. 
  * Python SDK: version 1.0 and higher.
  * Java SDK: As of Jan 30, 2018 has support as part of main source code, which is still not released. See [merged pull request](https://github.com/aws/aws-xray-sdk-java/pull/9).

X-Ray does provide support for configuring additional services (such as S3 and SNS) to include parameter info through a parameter whitelist configuration file. We've created separate projects to manage parameter whitelist configurations for the Java and Node X-Ray SDKs. For Python the same parameter whitelist file of Node can be used as specified below.

### X-Ray SDK for Node

The X-Ray SDK for Node provides a direct interface to append parameter whitelist configurations. Use our npm module to obtain additional parameter whitelist configurations which are not available as part of the X-Ray SDK.

Install the npm module:

```
npm install --save aws-xray-parameter-whitelist
```

Then during the initialization of your application add the following configuration code:

```javascript
const AWSXRay = require('aws-xray-sdk');
const whitelists = require('aws-xray-parameter-whitelist');
AWSXRay.captureAWS(require('aws-sdk')); //standard capture code of X-Ray to catch AWS SDK calls
AWSXRay.appendAWSWhitelist(whitelists.sns_whitelist);
```

More info available at the project page: https://github.com/functionalone/aws-xray-parameter-whitelist-node 

### X-Ray SDK for Python

The X-Ray SDK for Python doesn't expose a direct interface to append parameter whitelist configurations as in the X-Ray Node SDK. But there is an option to modify the whitelist object of the SDK directly as suggested in the following form post: https://forums.aws.amazon.com/message.jspa?messageID=802327#802327 (Note: the `whitelist` object has moved since the post, see the code below)  The Python SDK uses the same parameter whitelist syntax as used by the X-Ray Node SDK, with a slight difference in the way operations are named. Operations are required to start with an uppercase. Use the following procedure to obtain additional parameter whitelist configurations which are not available as part of the X-Ray SDK:

To add support for SNS, copy the sns_whitelist.json from: https://github.com/functionalone/aws-xray-parameter-whitelist-node/blob/master/resources/sns_whitelist.json 

Then during the initialization configure X-Ray to use the sns parameter whitelist. Example code (Note: you may need to modify the path to `sns_whitelist.json`) :

```python
from aws_xray_sdk.core import patch_all
from aws_xray_sdk.ext.boto_utils import whitelist

patch_all() # standard patch code to catch AWS SDK calls

# code to configure xray to provide parameter info for sns
with open('./sns_whitelist.json', 'r') as data_file:
    sns_whitelist = json.load(data_file)
operations = sns_whitelist['services']['sns']['operations']
for op in operations.keys():
    op_cap = op[:1].upper() + op[1:]
    operations[op_cap] = operations.pop(op)
whitelist['services']['sns'] = sns_whitelist['services']['sns']
```

### X-Ray SDK for Java

Use the `aws-xray-parameter-whitelist-instrumentor` jar which is a drop-in replacement for `aws-xray-recorder-sdk-aws-sdk-instrumentor`. It contains a pre-configured parameter whitelist file with additional support for S3. There is no need to modify source code to use this parameter whitelist configuration. There is only need to add the jar to the class path instead of the `aws-xray-recorder-sdk-aws-sdk-instrumentor` jar. It is possible to add the `aws-xray-parameter-whitelist-instrumentor` jar as a dependency via the JCenter repository. For example, if you are using Gradle you will need to add the following section to the repositories closure:

```
repositories {
    jcenter()
}
```

And then add the following compilation dependency:
```
compile 'com.github.functionalone:aws-xray-parameter-whitelist-instrumentor:<version>'
```

More info and further configuration options are available at the project page: https://github.com/functionalone/aws-xray-parameter-whitelist-java

# Usage and command line options

To list all the options run:

`xray-privilege-scan --help`

All command line options are optional:

```
  Usage: xray-privilege-scan [options]


  Options:

    -V, --version                 output the version number
    -s, --start-time <timestamp>  Start time as Unix timestamp (seconds since 1970-01-01 00:00:00 UTC). If left out will use: (current time - time range).
    -r, --time-range <minutes>    Time range in minutes to scan from start time. (default: 60)
    -c, --compare                 Compare current role and generated roles. Output a json report.
    -v, --verbose                 Output verbose logs to the console (info and above).
    -f, --filter <expression>     Filter expression to use when scanning xray. See AWS docs for synatx.
    -h, --help                    output usage information
```

Example usage to scan last 12 hours of X-Ray traces with verbose output:

```
xray-privilege-scan -v -r 720
```

Once completed `xray-privilege-scan` will generate a policy document per Lambda Function. Each document will be named with a unique id of the form:

`0fcacc5b207ffcd533267d2020962975.policy.json`

Additionally it will printout a summary to standard out specifying the function names and documents generated. 

The policy document is a json document conforming to the AWS Policy language with an additional field of: `Description`. The `Description` field will contain the AWS ARN of the Lambda Function this policy is for. The `Description` field is not part of the AWS Policy language and should be removed if copying the policy to AWS IAM.

## Supported AWS Services

The following services are supported as part of an X-Ray scan (unsupported services will be ignored during a scan):

* DynamoDB
* S3
* SQS
* Lambda
* SNS

If you would like to see additional services, please let us know by openning an [issue](https://github.com/functionalone/aws-least-privilege/issues).

## Compare Mode

To use compare mode specify the `-c` command line option. With compare mode the tool will check what role is currently attached to a Lambda Function. It will then compare the role's permissions against the permissions discovered through the X-Ray scan. If any of the roles contain excessive permissions as compared to the permissions actually discovered in the X-Ray scan, an excessive permissions report in json format will be generated. The file will be named: `excessive_permissions.json`. The report will include an entry for each Lambda Function with the permissions found as excessive. Sample report:

```json
[
  {
    "arn": "arn:aws:lambda:us-east-1:123456789:function:notes-app-api-prod-delete",
    "role": "arn:aws:iam::123456789:role/notes-app-api-prod-delete-us-east-1-lambdaRole",
    "excessPermissions": [
      {
        "Action": [
          "dynamodb:PutItem"
        ],
        "Resource": "arn:aws:dynamodb:us-east-1:*:table/notes",
        "Effect": "Allow"
      }
    ]
  }
]
```

**Note:** The comparison currenlty takes into account only "Allow" statements. If the role contains a policy with "Deny" statements these will not be considered for the comparison.

## More Info

**Detailed walk through with a sample application:** [Using AWS X-Ray to achieve Least Privilege Permissions](https://medium.com/@glicht/using-aws-x-ray-to-achieve-least-privilege-permissions-93dfd6701318)

**AWS X-Ray Filter Expression Documentation:** https://docs.aws.amazon.com/xray/latest/devguide/xray-console-filters.html#console-filters-syntax

[npm-image]:https://img.shields.io/npm/v/aws-least-privilege.svg
[npm-url]:http://npmjs.org/package/aws-least-privilege
[sls-image]:http://public.serverless.com/badges/v3.svg
[sls-url]:http://www.serverless.com
[travis-image]:https://travis-ci.org/functionalone/aws-least-privilege.svg?branch=master
[travis-url]:https://travis-ci.org/functionalone/aws-least-privilege
[david-image]:https://david-dm.org/functionalone/aws-least-privilege/status.svg
[david-url]:https://david-dm.org/functionalone/aws-least-privilege
[coveralls-image]:https://coveralls.io/repos/github/functionalone/aws-least-privilege/badge.svg?branch=master
[coveralls-url]:https://coveralls.io/github/functionalone/aws-least-privilege?branch=master

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.12"></a>
## [0.1.12](https://github.com/functionalone/aws-least-privilege/compare/v0.1.11...v0.1.12) (2018-09-02)


### Features

* update dependencies ([4c2fe0b](https://github.com/functionalone/aws-least-privilege/commit/4c2fe0b))



<a name="0.1.11"></a>
## [0.1.11](https://github.com/functionalone/aws-least-privilege/compare/v0.1.10...v0.1.11) (2018-06-05)


### Features

* support travis build and coverage reporting ([fc82935](https://github.com/functionalone/aws-least-privilege/commit/fc82935))
* update dependencies ([08838ce](https://github.com/functionalone/aws-least-privilege/commit/08838ce))



<a name="0.1.10"></a>
## [0.1.10](https://github.com/functionalone/aws-least-privilege/compare/v0.1.9...v0.1.10) (2018-04-26)


### Features

* add aditional note about X-Ray configuration when generated policy is empty ([feebd1f](https://github.com/functionalone/aws-least-privilege/commit/feebd1f))



<a name="0.1.9"></a>
## [0.1.9](https://github.com/functionalone/aws-least-privilege/compare/v0.1.8...v0.1.9) (2018-03-20)


### Bug Fixes

* fixed issue where subsegments were not parsed when there was a custom local segment ([6632d7c](https://github.com/functionalone/aws-least-privilege/commit/6632d7c))



<a name="0.1.8"></a>
## [0.1.8](https://github.com/functionalone/aws-least-privilege/compare/v0.1.7...v0.1.8) (2018-03-12)


### Features

* support for sns (issue [#1](https://github.com/functionalone/aws-least-privilege/issues/1)) ([ee16055](https://github.com/functionalone/aws-least-privilege/commit/ee16055))



<a name="0.1.7"></a>
## [0.1.7](https://github.com/functionalone/aws-least-privilege/compare/v0.1.6...v0.1.7) (2018-02-21)


### Features

* support passing filter expression to use when scanning xray traces ([2d304c0](https://github.com/functionalone/aws-least-privilege/commit/2d304c0))



<a name="0.1.6"></a>
## [0.1.6](https://github.com/functionalone/aws-least-privilege/compare/v0.1.5...v0.1.6) (2018-02-06)


### Features

* Support new "compare mode". Allows comparing the X-Ray discovered permissions against the current permissions assigned to a Lambda Function. ([335b116](https://github.com/functionalone/aws-least-privilege/commit/335b116))



<a name="0.1.5"></a>
## [0.1.5](https://github.com/functionalone/aws-least-privilege/compare/v0.1.4...v0.1.5) (2018-01-21)


### Features

* support for extracting Lambda actions from traces ([a178dfb](https://github.com/functionalone/aws-least-privilege/commit/a178dfb))
* support for extracting SQS actions from traces ([6a109e7](https://github.com/functionalone/aws-least-privilege/commit/6a109e7))


src/lib/xray-trace-fetcher.ts
import { logger } from './logger';
import AWS from 'aws-sdk';
import _ from 'lodash';
import { SUPPORTED_SERVICES, AWSSegmentParsers } from './aws-segments/index';
import BBPromise from 'bluebird';
import { IamPolicyDocument, IamStatement } from './aws-segments/iam-policy';
import crypto from 'crypto';
import fs from 'fs';
import {Trace} from 'aws-sdk/clients/xray';
import { IamPolicyUtils } from './iam-policy-utils';

const DEF_TIME_RANGE = 60;
export const EXCESSIVE_PERMISSION_FILE = 'excessive_permissions.json';

export interface ScanXrayTracesConf {
  /**
   * start time to search for. If not specified will do a "last x min search"
   */
  startTime?: Date;
  /**
   * Time in minutes range to search. If startTime is not specified this will be used as a last X min search. 
   * Defaults to: 60
   */
  timeRangeMinutes?: number;
  /**
   * Filter expression as documented at: https://docs.aws.amazon.com/xray/latest/devguide/xray-console-filters.html#console-filters-syntax
   */
  filterExpression?: string;
  /**
   * Perform comparison against the application role and generate an excess permissions report
   */
  compareExistingRole?: boolean;
}

/**
 * Fetch xray traces. Combines the two functions of getTraceSummaries and batchGetTraces including support for pagination.
 */
export async function getXrayTraces(conf?: ScanXrayTracesConf) {
  try {
    logger.info('getXrayTraces: params: ', conf);
    const xray = new AWS.XRay();
    const timeRangeMS = ((conf && conf.timeRangeMinutes) || DEF_TIME_RANGE) * 60 * 1000;
    const StartTime = (conf && conf.startTime) || new Date(Date.now() - timeRangeMS);
    const EndTime = new Date(StartTime.getTime() + timeRangeMS);
    logger.info("GetTraceSummariesRequest: StarTime: %s, EndTime: %s", StartTime, EndTime);
    const summariesParams: AWS.XRay.Types.GetTraceSummariesRequest = {
      StartTime,
      EndTime,
      FilterExpression: (conf && conf.filterExpression) || undefined,
    };
    let paginationNum = 0;
    let nextToken: string | undefined;
    let summaries: AWS.XRay.TraceSummary[] = [];
    do {
      summariesParams.NextToken = nextToken;
      const traceSummariesResp = await xray.getTraceSummaries(summariesParams).promise();
      logger.debug('getTraceSummaries [token: %s] result: ', nextToken, traceSummariesResp);
      if(traceSummariesResp.TraceSummaries) {
        summaries = summaries.concat(traceSummariesResp.TraceSummaries);
      }
      nextToken = traceSummariesResp.NextToken;
      paginationNum++;
    } while (nextToken);    
    logger.info('getTraceSummaries total summaries after [%s] pagination: ', paginationNum, summaries.length);
    let batchNum = 0;
    let ids: string[] = [];
    // tslint:disable-next-line:max-line-length
    const batchResults = await BBPromise.map<AWS.XRay.TraceSummary, AWS.XRay.BatchGetTracesResult | null>(summaries, (trace, index, arrayLength) => {
      if (trace.Id) {
        ids.push(trace.Id);
      }
      //aws enforces max 5 ids in a batch. We do the request when we reach 5 or on the last index
      if (ids.length === 5 || index === (arrayLength - 1)) {
        logger.debug('Performing batchGetTraces with ids: ', ids);
        const res = xray.batchGetTraces({ TraceIds: ids }).promise();
        ids = [];
        batchNum++;
        return res;
      }
      return null;
    }, { concurrency: 10 });
    let traces: Trace[] = [];
    for (const res of batchResults) {
      if (res) {
        logger.debug("batchGetTraces result: ", res);
        if (!_.isEmpty(res.Traces)) {
          logger.debug('batchGetTraces count: ', res.Traces!.length);
          traces = traces.concat(res.Traces!);
        }
        else {
          logger.error('Got empty result from batchGetTraces for ids: ', ids);
        }
      }
    }
    logger.info('getXrayTraces total traces returned: %s, number of batches: %s', traces.length, batchNum);
    return traces;
  } catch (error) {
    logger.error("Failed fetching traces: ", error);
    throw error;
  }
}

/**
 * Map that maps a resource to set of actions
 */
export class ResourceActionMap extends Map<string, Set<string>> {
  constructor() {
    super();
  }

  addActionToResource(actionOp: string, resourceArn: string) {
    let actions = super.get(resourceArn);
    if(!actions) {
      actions = new Set();
      super.set(resourceArn, actions);    
    }
    actions.add(actionOp);
  }
}

/**
 * Map which maps a function to the set of ResourceActionMap
 */
export type FunctionToActionsMap = Map<string, ResourceActionMap>;

function parseSegmentDoc(doc: any, functionMap: FunctionToActionsMap, parentActionsMap?: ResourceActionMap, functionArn?: string) {
  //if this is a subsegment we will want to add to the parentActionsMap if this relates to an aws service
  if (parentActionsMap && doc.namespace === 'aws' && SUPPORTED_SERVICES.indexOf(doc.name) >= 0) {
    AWSSegmentParsers[doc.name](doc, parentActionsMap, functionArn!);
  }
  //see if this is a Lambda function. If so it will have aws.function_arn
  const arn = _.get(doc, 'aws.function_arn');
  let actionsMap: ResourceActionMap | undefined = parentActionsMap;
  if (arn) {
    logger.debug('Found arn: %s for Segment: %s', arn, doc.id);
    actionsMap = functionMap.get(arn);
    if (!actionsMap) {
      actionsMap = new ResourceActionMap();
      functionMap.set(arn, actionsMap);
    }
  }
  if (!_.isEmpty(doc.subsegments)) {
    for (const subsegment of doc.subsegments) {
      parseSegmentDoc(subsegment, functionMap, actionsMap, arn);
    }
  }
}

/**
 * Parse an xray trace return the passed in functionMap
 * @param trace 
 * @param functionMap will populate with the found results
 */
export function parseXrayTrace(trace: AWS.XRay.Trace, functionMap: FunctionToActionsMap): FunctionToActionsMap {
  if (_.isEmpty(trace.Segments)) {
    logger.info('No segments found for TraceId: ', trace.Id);
    return functionMap;
  }
  //we search for lambda segments and then follow into sub segments
  for (const segment of trace.Segments!) {
    const docStr = segment.Document;
    if (_.isEmpty(docStr)) {
      logger.warn("Got segment [id: %s] with empty Document.", segment.Id);
    }
    else {
      parseSegmentDoc(JSON.parse(docStr!), functionMap);
    }
  }
  return functionMap;
}

export function parseXrayTraces(traces: AWS.XRay.Trace[], functionMap: FunctionToActionsMap): FunctionToActionsMap {
  if (_.isEmpty(traces)) {
    return functionMap;
  }
  for (const trace of traces) {
    parseXrayTrace(trace, functionMap);
  }
  return functionMap;
}

export async function getFunctionActionMapFromXray(conf?: ScanXrayTracesConf) {
  const traces = await getXrayTraces(conf);
  const map: FunctionToActionsMap = new Map();
  return parseXrayTraces(traces, map);
}

export function createIAMPolicyDoc(map: ResourceActionMap, functionArn: string) {
  const doc: IamPolicyDocument = {
    Version: "2012-10-17",
    Description: "Generated policy from xray scan for: " + functionArn,
    Statement: [],
  };
  //consolidate resources with the same functions into one entry
  const actionMap = new Map<string, {resources: string[], actions: string[]}>();
  for (const entry of map.entries()) {
    const match = entry[0].match(/arn:aws:(.+?):/);
    let service = '';
    if(!match || match.length < 2) {
      logger.error("Couldn't extract service name from resource arn: ", entry[0]);
    }
    else {
      service = match[1];
    }
    const actions = Array.from(entry[1], (s) => `${service}:${s}`).sort();
    const actionsKey = actions.join();
    let actionMapEntry = actionMap.get(actionsKey);
    if(!actionMapEntry) {
      actionMapEntry = {
        resources: [],
        actions,
      };
      actionMap.set(actionsKey, actionMapEntry);
    }
    actionMapEntry.resources.push(entry[0]);
  }
  for (const val of actionMap.values()) {
    const stm: IamStatement = {
      Effect: "Allow",      
      Action: val.actions,
      Resource: val.resources,
    };
    doc.Statement!.push(stm);  
  }
  if(_.isEmpty(doc.Statement)) {
    //append more details in the description
    // tslint:disable-next-line:max-line-length
    doc.Description += ". Note: No access to AWS resources were found for this function. It may be that X-Ray wasn't configured correctly. See: https://github.com/functionalone/aws-least-privilege#x-ray-setup on how to setup X-Ray.";
  }
  return doc;
}

export interface AppExcessPermissions {
  arn: string;
  role: string;
  excessPermissions: IamStatement[];
}

export async function compareLambdaRoleToPolicy(lambdaArn: string, policy: IamPolicyDocument) {
  const iamUtils = new IamPolicyUtils();
  if(!policy.Statement) {
    return;
  }
  const lambda = new AWS.Lambda();
  const config = await lambda.getFunctionConfiguration({FunctionName: lambdaArn}).promise();
  if(config.Role)
  {
    const roleName = config.Role.substr(config.Role.lastIndexOf('/') + 1);
    const roleStatements = await iamUtils.getCombinedStatementsForRole(roleName);
    const excessStatements = iamUtils.checkExcessPolicyPermissions(roleStatements, policy.Statement);
    if(!_.isEmpty(excessStatements)) {
      const res: AppExcessPermissions = {
        arn: lambdaArn,
        role: config.Role,
        excessPermissions: excessStatements!,
      };
      return res;
    }
  }  
}

export interface GeneratedPolicy {
  Arn: string;
  FileName?: string;
  Policy: IamPolicyDocument;
}

export interface ScanResult {
  GeneratedPolicies: GeneratedPolicy[];
  ExcessPermissions: AppExcessPermissions[];
}

/**
 * Will scan xray and return a ScanResult
 * @param conf 
 */
export async function scanXray(conf: ScanXrayTracesConf): Promise<ScanResult> {
  const map: FunctionToActionsMap = await getFunctionActionMapFromXray(conf);  
  const res: ScanResult = {
    GeneratedPolicies: [],
    ExcessPermissions: [],
  };
  const promiseArr: Promise<AppExcessPermissions|undefined>[] = [];
  for (const entry of map.entries()) {
    const policyDoc = createIAMPolicyDoc(entry[1], entry[0]);
    res.GeneratedPolicies.push({
      Arn: entry[0],
      Policy: policyDoc,
    });
    if(conf.compareExistingRole) {
      promiseArr.push(compareLambdaRoleToPolicy(entry[0], policyDoc));
    }    
  }
  if(conf.compareExistingRole) {
    const permissionsRes = await Promise.all(promiseArr);
    for (const perm of permissionsRes) {
      if(perm) {
        res.ExcessPermissions.push(perm);
      }
    }    
  }
  return res;
}

/**
 * 
 * @param conf 
 * @return object mapping function arn to file name
 */
export async function scanXrayAndSaveFiles(conf: ScanXrayTracesConf) {
  const res = await scanXray(conf);  
  for (const generatedPolicy of res.GeneratedPolicies) {    
    const hash = crypto.createHash('md5');
    hash.update(generatedPolicy.Arn);
    const fileName = hash.digest('hex') + ".policy.json";
    generatedPolicy.FileName = fileName;
    try {
      fs.writeFileSync(fileName, JSON.stringify(generatedPolicy.Policy, undefined, 2));
    } catch (error) {
      logger.error("Failed writing to file: [%s] ",fileName, error);
    }
  }
  if(!_.isEmpty(res.ExcessPermissions)) {
    fs.writeFileSync(EXCESSIVE_PERMISSION_FILE,  JSON.stringify(res.ExcessPermissions, undefined, 2));
  }
  return res;
}

// export XRay.Trace;
